scoring/score.py

```diff
@@ -13,21 +24,158 @@ def __init__(self, message: str, *, code: str) -> None:
+                code='invalid_districts',
+            )
+
+        bad_highest = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            if highest and highest not in ZONE_COLOURS:
+                bad_highest[name] = highest
+        if bad_highest:
+            raise InvalidScoresheetException(
+                f"Invalid pallets specified as the highest in some districts -- "
+                f"must be a single pallet from "
+                f"{join_or(repr(x) for x in ZONE_COLOURS)}.\n"
+                f"{bad_highest!r}",
+                code='invalid_highest_pallet',
+            )
+
+        bad_pallets = {}
+        for name, district in self._districts.items():
+            extra = district['pallet_counts'].keys() - ZONE_COLOURS
+            if extra:
+                bad_pallets[name] = extra
+        if bad_pallets:
+            raise InvalidScoresheetException(
+                f"Invalid pallets specified in some districts -- must be from "
+                f"{join_or(repr(x) for x in ZONE_COLOURS)}.\n"
+                f"{bad_pallets!r}",
+                code='invalid_pallets',
+            )
+
+        bad_highest2 = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            if highest and highest not in district['pallet_counts']:
+                bad_highest2[name] = (highest, district['pallet_counts'].keys())
+        if bad_highest2:
+            detail = "\n".join(
+                (
+                    (
+                        f"District {name} has only {join_and(pallets)} so "
+                        f"{highest!r} cannot be the highest."
+                    )
+                    if pallets
+                    else (
+                        f"District {name} has no pallets so {highest!r} cannot "
+                        "be the highest."
+                    )
+                )
+                for name, (highest, pallets) in bad_highest2.items()
+            )
+            raise InvalidScoresheetException(
+                f"Impossible pallets specified as the highest in some districts "
+                f"-- must be a pallet which is present in the district.\n"
+                f"{detail}",
+                code='impossible_highest_pallet',
+            )
+
+        missing_highest = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            pallet_counts = district['pallet_counts']
+            if len(pallet_counts) == 1 and not highest:
+                pallet, = pallet_counts.keys()
+                missing_highest[name] = pallet
+        if missing_highest:
+            raise InvalidScoresheetException(
+                f"Some districts with pallets from a single team are missing "
+                "specification of the highest.\n"
+                f"{missing_highest!r}",
+                code='missing_highest_pallet',
+            )
```

Comment:
Is this checking that if there's at least one pallet in the scoring zone that there should be one listed in highest? Because there are scenarios where there is only one pallet in the scoring zone and it does not count as being highest (e.g. if it's on the floor)

Reply:
This is checking the case where there's pallets from _exactly one_ team present in the scoring zone, in which case (I believe?) then that team's pallets _must_ be the highest.

Reply:
Ah okay. I think in the case of only one pallet on the floor, no pallet is highest as pallets must not be touching the floor to count as highest, so no pallets would be doubled in that scenario

Reply:
:facepalm: d'oh, forgot about that rule. Yeah, we can remove this check then and I thiiink everything else still holds?

Reply:
Yeah I think we can remove it. There's a consideration here for how marshals will actually end up doing the highest - I can see an argument for telling them to write a single line to indicate no highest to show that they have checked, but I suppose score entry can just treat these as a blank square when actually entering them

Reply:
0cc5b57c2483655ec678f53395fa85d460b7f379 removes that check. 4ca07e188f0fe0f9a9149fe5975e2f479eec5a6b introduces the inverse check, only for the outer districts.